core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -330,54 +322,51 @@ private object RTS {
 
                     val value = io.value()
 
-                    curIo = nextInstr[E](value, stack)
-
-                    if (curIo eq null) {
-                      result = ExitResult.succeeded(value)
-                    }
+                    curIo = nextInstr(value)
 
                   case IO.Tags.Strict =>
                     val io = curIo.asInstanceOf[IO.Strict[Any]]
 
                     val value = io.value
 
-                    curIo = nextInstr[E](value, stack)
-
-                    if (curIo eq null) {
-                      result = ExitResult.succeeded(value)
-                    }
+                    curIo = nextInstr(value)
 
                   case IO.Tags.SyncEffect =>
                     val io = curIo.asInstanceOf[IO.SyncEffect[Any]]
 
                     val value = io.effect()
 
-                    curIo = nextInstr[E](value, stack)
-
-                    if (curIo eq null) {
-                      result = ExitResult.succeeded(value)
-                    }
+                    curIo = nextInstr(value)
 
                   case IO.Tags.AsyncEffect =>
                     val io = curIo.asInstanceOf[IO.AsyncEffect[E, Any]]
 
-                    val cancel = enterAsync()
-
-                    io.register(resumeAsync) match {
-                      case Async.Now(value) =>
-                        // Value returned synchronously, callback will never be
-                        // invoked. Attempt resumption now:
-                        if (shouldResumeAsync()) {
-                          curIo = IO.done(value)
-                        } else {
-                          curIo = null
-                        }
+                    // By default, we'll resume asynchronously, so this
+                    // evaluation loop should halt:
+                    curIo = null
 
-                      case Async.MaybeLater(cancel0) =>
-                        // Store the canceler:
-                        cancel.set(cancel0)
+                    // Enter suspended state:
+                    val cancel = enterAsync()
 
-                        curIo = null
+                    if (cancel != null) {
+                      // It's possible we were interrupted prior to entering
+                      // suspended state. So we have to check that condition,
+                      // and if so, do not initiate the async effect (because
+                      // otherwise, it would not be interrupted).
+                      try {
+                        if (state.get.interrupted && noInterrupt == 0) curIo = IO.interrupt
```

Comment:
`if(shouldDie)`?

Reply:
`shouldDie` is cached in each thread, so it's stale (however, it's much faster to access than `state.get.interrupted`). So here we have to use the slow method to make sure we evade race conditions between the start of an async action and a kill attempt.